File: app/src/main/java/io/mgba/Activities/SettingsCategoriesActivity.java
package io.mgba.Activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.LinkedList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.mgba.Adapters.SettingsCategoriesAdapter;
import io.mgba.Constants;
import io.mgba.Data.Settings.SettingsCategory;
import io.mgba.R;

public class SettingsCategoriesActivity extends AppCompatActivity {

    @BindView(R.id.recyclerview)
    RecyclerView recyclerview;
    private List<SettingsCategory> settings;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings_categories);
        ButterKnife.bind(this);

        settings = new LinkedList<>();
        settings.add(new SettingsCategory(getResources().getString(R.string.settings_audio), R.drawable.ic_audiotrack_black_24dp));
        settings.add(new SettingsCategory(getResources().getString(R.string.settings_video), R.drawable.ic_personal_video_black_24dp));
        settings.add(new SettingsCategory(getResources().getString(R.string.settings_emulation), R.drawable.ic_gamepad_black_24dp));
        settings.add(new SettingsCategory(getResources().getString(R.string.settings_bios), R.drawable.ic_memory_black_24dp));
        settings.add(new SettingsCategory(getResources().getString(R.string.settings_paths), R.drawable.ic_folder_black_24dp));
        settings.add(new SettingsCategory(getResources().getString(R.string.settings_customization), R.drawable.ic_palette_black_24dp));

        setupToolbar();
        setupRecyclerView();
    }

    private void setupRecyclerView() {
        recyclerview.setHasFixedSize(true);
        recyclerview.setLayoutManager(new LinearLayoutManager(this));
        recyclerview.setAdapter(new SettingsCategoriesAdapter(settings, this, (s) -> {
            Intent it = new Intent(getApplicationContext(), SettingsActivity.class);
            it.putExtra(Constants.ARG_SETTINGS_ID, s.getTitle());
            startActivity(it);
        }));
    }

    private void setupToolbar() {
        if(getSupportActionBar() != null){
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: app/src/main/java/io/mgba/settings/categories/SettingsCategoriesActivity.kt
package io.mgba.settings.categories

import androidx.recyclerview.widget.LinearLayoutManager
import io.mgba.R
import io.mgba.base.BaseActivity
import io.mgba.base.BaseAdapter
import io.mgba.data.settings.Settings
import io.mgba.settings.categories.recyclerView.SettingsAdapter
import kotlinx.android.synthetic.main.activity_settings.*

class SettingsCategoriesActivity : BaseActivity<SettingsCategoriesViewModel>(){
    private val settings: List<Settings> = listOf(Settings(getString(R.string.settings_audio), R.drawable.ic_audiotrack),
                                                  Settings(getString(R.string.settings_video), R.drawable.ic_screen),
                                                  Settings(getString(R.string.settings_emulation), R.drawable.ic_gamepad),
                                                  Settings(getString(R.string.settings_bios), R.drawable.ic_memory),
                                                  Settings(getString(R.string.settings_paths), R.drawable.ic_folder),
                                                  Settings(getString(R.string.settings_customization), R.drawable.ic_themes))

    override fun getLayout(): Int = R.layout.activity_settings
    override fun getViewModel(): Class<SettingsCategoriesViewModel> = SettingsCategoriesViewModel::class.java

    override fun onCreate() {
        setupToolbar()
        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        list.setHasFixedSize(true)
        list.layoutManager = LinearLayoutManager(this)
        list.adapter = SettingsAdapter(this) { s -> vm.onClick(s) }
        (list.adapter as BaseAdapter<Settings>).data = ArrayList(settings)
    }

    private fun setupToolbar() {
        supportActionBar?.let {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            supportActionBar!!.setDisplayShowHomeEnabled(true)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }
}
